libbeat/publisher/pipeline/output.go

```diff
@@ -71,56 +70,51 @@ func (w *clientWorker) run() {
 }
 
 func (w *netClientWorker) Close() error {
-	w.closed.Store(true)
+	close(w.done)
 	return w.client.Close()
 }
 
 func (w *netClientWorker) run() {
-	for !w.closed.Load() {
-		reconnectAttempts := 0
-
-		// start initial connect loop from first batch, but return
-		// batch to pipeline for other outputs to catch up while we're trying to connect
-		for batch := range w.qu {
-			batch.Cancelled()
-
-			if w.closed.Load() {
-				logp.Info("Closed connection to %v", w.client)
-				return
-			}
-
-			if reconnectAttempts > 0 {
-				logp.Info("Attempting to reconnect to %v with %d reconnect attempt(s)", w.client, reconnectAttempts)
-			} else {
-				logp.Info("Connecting to %v", w.client)
-			}
-
-			err := w.client.Connect()
-			if err != nil {
-				logp.Err("Failed to connect to %v: %v", w.client, err)
-				reconnectAttempts++
+	var (
+		connected         bool
+		reconnectAttempts int
+	)
+
+	for {
+		select {
+		case <-w.done:
+			logp.Info("Closed connection to %v", w.client)
+			return
```

Comment:
This looks like a potential race. When calling `select` the runtime chooses the channel order by random. That is we can have this sequence of events happening:

```
1. close(w.done)
2. batch := <-w.qu
3. w.client.Close()
4. batch.Cancelled() // connection was already closed before
5. w.client.Connect() // reconnect -> fd leak
6. <- w.done
7. return
```

alternatively:

```
1. batch := <-w.qu
2. close(w.done)
3. w.client.Close()
4. w.client.Publish() -> error -> connection close
5. batch := <-w.qu
6. batch.Cancelled()
7. w.client.Connect() // fd leak
8. <-w.done
9. return
```

This is why we used to have the boolean, so to check if the output has been closed or if we can reconnect. The client must have no chance to reconnect after the output has been closed.

Reply:
After talking with @urso a mix of both a boolean flag and a done channel would be something to change.

Reply:
Looking at the implementation we need to change a bit the behavior of close, close need to be done in sync instead of async because we want to be able to return the error that could be potentially raised by a client when closing.

Also, looking at the controller when a reload happen we do not actually log the error on close we just silently ignore it.


We need something like this:
````
1. batch := <-w.qu
2. close(w.done)
3. Get out of the loop.
4. Wait until the loop is complete
9. w.client.Close()
9. return
```